### PR TITLE
Feature/exp components command

### LINKED ISSUE #9625: [mesheryctl] New command: `exp components [list|view|search]`
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

### Current Behavior
Create an `exp components [list|view|search]` command similar to that of `exp model` command.

---

### Contributor Resources

- ⌨️ [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](http://discuss.meshery.io) and [Layer5 Community Slack](https://slack.meshery.io)

### PR DESCRIPTION
**Pull request by RipulHandoo**
**Notes for Reviewers**

This PR fixes #9625 

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: mesheryctl/internal/cli/root/system/component.go [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"gopkg.in/yaml.v2"
+)
+
+// represents the `mesheryctl exp components list` command
+var listComponentCmd = &cobra.Command{
+	Use:   "list",
+	Short: "List componets",
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/component.go:26]:
```suggestion
	Short: "List registered components",
```

File: mesheryctl/internal/cli/root/system/component.go [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+	Long:  `List all the components of the system`,
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/component.go:27]:
```suggestion
	Long:  `List all components registered in Meshery Server`,
```

**Comment by saurabh100ni** [comment]
Thanks for raising PR @RipulHandoo, please add this in [Docs](https://github.com/meshery/meshery/pull/9716/files) as your item to discuss it in the Meshery Development Meet today at 8:30 pm IST.
Check out [Community Calendar ](https://layer5.io/community/calendar)for meet Details.

**Comment by leecalcote** [comment]
@RipulHandoo will you have a look at the failing lint checks?

**Comment by RipulHandoo** [comment]
@leecalcote  Sir, Currently, we are encountering a lint check failure attributed to missing fields (Category, Name, Version) in the ComponentDefinition used in our Model struct.
Please take a look at [this](https://github.com/meshery/meshery/pull/9726/files#diff-09fcaa361285a17c501b91b394c9c6ef119700f161e9ac10f7eb6572bde16336R105).
I would greatly appreciate your insights on whether it is advisable to incorporate these fields.

**Comment by l5io** [comment]
🚀 Preview for commit fc9b2ad4f2d6c28ce7a22654501a112acda541a0 at: https://658d511ec7fcc49f46c83ab9--meshery-docs-preview.netlify.app

**Comment by leecalcote** [comment]
@RipulHandoo watch out for accidentally intermixing commits between branches

**Comment by RipulHandoo** [comment]
> Hi @RipulHandoo since this is an exp command, I suggest you to add this in this entry in the file marked below...
> 
> https://github.com/meshery/meshery/blob/047800b0c42b9c100821f59601e11ac20d30b568/mesheryctl/internal/cli/root/experimental/experimental.go#L46-L49

**Comment by leecalcote** [comment]
@RipulHandoo are expecting these build failures?

**Comment by RipulHandoo** [comment]
@leecalcote Yes sir. I am expecting these failures. As some of the fields are not defined in the ComponentDefinition.

**Comment by leecalcote** [comment]
Thanks much for confirming, @RipulHandoo. Makes sense.

@Yashsharma1911 your 👀 on this would be helpful.

**Comment by leecalcote** [comment]
@RipulHandoo there's a spot with your name on it here - https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit

**Comment by leecalcote** [comment]
> @RipulHandoo there's a spot with your name on it here - https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit

Which you have now deleted. Ok, never mind.

**Comment by RipulHandoo** [comment]
> > @RipulHandoo there's a spot with your name on it here - https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit
> 
> Which you have now deleted. Ok, never mind.

@leecalcote Sir. 
Unfortunately, I won't be able to provide a live demonstration of the new Mesheryctl commands during today's meeting as I am currently traveling and away from my keyboard. 
However, if it's feasible, I can still join the meeting via my mobile device to discuss any updates or answer questions.

**Review by leecalcote — COMMENTED** [review]
@RipulHandoo please generate the model, component, relationship, and so on structs using the OpenAPI generator.

I searched Meshery Docs, but @MUzairS15 I couldn't find these instructions. Please point to the guide that has the schema-driven instructions in Meshery Docs.

**Comment by leecalcote** [comment]
@RipulHandoo here's a quick reference - https://docs.google.com/presentation/d/1SQMfyu5shjpGKlYONdVzOtd7UYTgLWBcgUvHMLCZ2tY/edit#slide=id.g232ac6f646e_0_158

**Comment by leecalcote** [comment]
https://github.com/meshery/schemas/tree/master/schemas/meshmodel/schemas

**Comment by RipulHandoo** [comment]
> https://github.com/meshery/schemas/tree/master/schemas/meshmodel/schemas

This helped a lot in understanding the struct fields.
Thanks!!

**Comment by RipulHandoo** [comment]
@leecalcote sir, I've successfully tested the `mesheryctl exp components` commands locally, and they are functioning as expected. However,` golangci-lint` is encountering issues. I would greatly appreciate your guidance in resolving this matter. Thank you

**Comment by leecalcote** [comment]
Please demo this command at the next Meshery dev meeting.

**Comment by RipulHandoo** [comment]
> Please demo this command at the next Meshery dev meeting.

Sure Sir. 🚀

**Comment by alphaX86** [comment]
@RipulHandoo good job on demo. Current items you need to look now are as follows:
- UX message for invalid search. Eg: If a user entered a wrong/misspelled component, the command should print a message, not left void.
- Pagination stuff you did is good. But I feel something is missing there, so I'll look it up and tell what needs to be done.

I'll create a design spec doc in meantime to track whatever we're doing. You can involve that as well.

**Comment by RipulHandoo** [comment]
@alphaX86 Thanks for the feedback on the demo! I'll work on adding the UX message for invalid search. Also, I'll be waiting for your guidance on the pagination part. I'm ready to help with the design spec doc too. 
Thanks again!

**Comment by leecalcote** [comment]
@alphaX86 @RipulHandoo have either of you worked up a design spec?

**Comment by RipulHandoo** [comment]
> > @alphaX86 @RipulHandoo have either of you worked up a design spec?
> 
> A doc is made. I'll work with @RipulHandoo as I'm free tomorrow mostly.

Works fine with me. Thanks for the update.

**Comment by alphaX86** [comment]
Design spec doc link: https://docs.google.com/document/d/1pxBDl0qgA6RmfH2fm_9j9Cp556QDZaqPaAtPIY3CHX4/edit?usp=sharing

**Comment by RipulHandoo** [comment]
Hi @leecalcote and @alphaX86,

Could you please review the failing Mesheryctl/golangci-lint test in relation to the latest PR? I'm considering closing the current PR and opening a new one since the tests are failing due to some undefined fields in [docs_test.go](https://github.com/meshery/meshery/pull/9726/files). Notably, this file hasn't been modified in this PR.

Your assistance in resolving this matter is greatly appreciated.

**Comment by alphaX86** [comment]
> Hi @leecalcote and @alphaX86,
> 
> Could you please review the failing Mesheryctl/golangci-lint test in relation to the latest PR? I'm considering closing the current PR and opening a new one since the tests are failing due to some undefined fields in [docs_test.go](https://github.com/meshery/meshery/pull/9726/files). Notably, this file hasn't been modified in this PR.
> 
> Your assistance in resolving this matter is greatly appreciated.
> 

I'll look on the golangci-lint and the unit test issue. For now, implement the changes I've requested previously. If all fails up even after fixes etc, then we might consider a new PR. Until then, I suggest you to continue.

**Comment by RipulHandoo** [comment]
@leecalcote @alphaX86, should I add pagination to the `mesheryctl exp components list` command?

https://github.com/meshery/meshery/assets/107461226/da009d2e-6f3c-4f71-9ebb-05774c9a7124

**Comment by leecalcote** [comment]
@alphaX86 were you able to figure this out?

**Comment by leecalcote** [comment]
@RipulHandoo after you address this final merge conflict, let's merge.

**Pull request merged**